### DISCUSSION TITLE
Any idea to check similar images?

### DISCUSSION THREAD

**Discussion by Elaphe-GitHub** [ImageMagick/ImageMagick#5196] [Consulting]
I have a collection of nearly 100.000 downloaded images. I add new images to this collection very often, using ACDSee to browse and view the files and an AutoHotkey script to copy the new images. When a file with the same name is found in the target location, my script compares the file sizes. If they are not the same, that means that both images are different, so the file is copied with a random suffix added to its name. I have an alternative method which reads the md5 checksum of the image to be copied and checks if it is listed in the checksum database of all my images (everytime a file is copied, its checksum is added to this database). Therefore, I can also check if an image is already in my collection even if they have different names. The problem is with images that have different names and different sizes, such as photos which are the same but have different resolution or compression. Can ImageMagick help my script with this? Any idea? I know I can check image duplicates with Visipics (something that I do do my collection every now and then), but my idea is to this on the fly. Thankyou.

**Comment by snibgo** [comment]
Files often contain metadata for the date of the file, so md5 checksum won't help.

IM can compare images that are the same size (in pixels), for example:
```
magick compare -metric RMSE a1.png a2.png NULL:
```
That gives a number between 0.0 (the inputs are identical) to 1.0 (they are totally different).

If the inputs are the same but with different lossy compression, the number will be small. As a rule of thumb, a number less than 0.01 means the difference won't be visually obvious.

**Comment by fmw42** [comment]
You could use perceptual hash for that. 

magick identify -verbose image.suffix

will provide the 42 perceptual hash floating point values.

See http://www.fmwconcepts.com/misc_tests/perceptual_hash_test_results_510/index.html

What is your Imagemagick version and platform?

**Comment by Elaphe-GitHub** [comment]
That is interesting. It is very useful to compare 2 images that have slight differences. However, I can't see how it could help me with my problem, since I do not have a known target file to compare the new file to be copied. My script would have to compare the new file with every single file in my collection, and that would take ages.

That idea of creating a perceptual hash would be the only alternative that seems to be useful. That would mean creating hashes for my 100.000 files and store the info to a text file. MD5 is already slow, but it seems this method is 10 times slower.

I don't see an easy solution. I would be happy if I just could find a way to make my script identify the same image in case it has a different size (for instance due to different level of compression. That is the most frequent case of duplicates I have found.

**Comment by snibgo** [comment]
There are many possibilities, depending on the nature of your images. For example, you might base the hash on some normalisation of the images, for example resized to fit in a 500x500 box.

You might devise your own simplified hash. For example, the overall mean of all the channels. If the means of two images are identical (or very close) then the images are *probably* identical (or very similar). So only such pairs need the more expensive test of "magick compare -metric RMSE a1.png a2.png NULL:".

**Comment by Elaphe-GitHub** [comment]
Thank you. Can you provide an example of that kind of simplified hash? Remember that I have to run it over 100.000 images to create the database, and I hope my computer to finish the task before Christmas, if possible. The goal is to spot different .jpg compresions of images that are visually identical.

**Comment by fmw42** [comment]
What is your Imagemagick version and OS?  I have some other simpler perceptual hash scripts for Unix-like systems (Linux, Mac OSX or Windows 10 with Unix). See http://www.fmwconcepts.com/imagemagick/phashes/index.php

**Reply by Elaphe-GitHub** [reply]
Imagemagick 7.0.8 for Windows 10.

**Comment by snibgo** [comment]
> Can you provide an example of that kind of simplified hash?

I did: the overall mean of all the channels. This would be a poor hash of processed sRGB photographs, which often have a mean close to 0.5. Standard deviation may be better.

> Remember that I have to run it over 100.000 images to create the database, and I hope my computer to finish the task before Christmas, if possible.

If the process takes 1 second/image, the corpus would take a little over one day.
